### PR TITLE
replaced watch with watchFile

### LINKED ISSUE #9781: Jupyter Books viewlet shows duplicated book structure when TOC is updated
Azure Data Studio version: azuredatastudio-insiders 1.17.0-insider (be83b31e378774ed14879fb1889c47422e68bea5, 2020-03-27T06:21:06.051Z)
OS version: Windows_NT x64 10.0.18363

Issue Type: <b>Bug</b>

1. Update toc.yml with another line item.
2. Press Ctrl + S to save.
3. The Saved Books subviewlet in Jupyter Books viewlet will show duplicated books. 

JupyterBookViewletTOCUpdate

image

### PR DESCRIPTION
**Pull request by MaddyDev**
Issue: Jupyter Books viewlet shows duplicated book structure when TOC is updated.

Found out that this is due to the fs.watch api with the below issues

- fs.watch is [unstable](https://nodejs.org/docs/latest/api/fs.html#fs_caveats)

- has [known "behaviour"](https://github.com/nodejs/node-v0.x-archive/search?q=issues+fs.watch+twice&ref=cmdform&type=Issues) with regards to repeated notifications. Specifically, the [windows case ](https://github.com/nodejs/node-v0.x-archive/issues/1970#issuecomment-2599352) being a result of windows design, where a single file modification can be multiple calls to the windows API

Though fs.watchFile is also unstable, this seems to fix the issue and work as expected.

**Note**: if you're editing toc file in ADS, please note that every pause auto-saves the file and triggers the watcher event; which tries to refresh the tree with invalid toc.yml and hence might end up with error notifications. But once you've completed your update (with valid syntax), the tree will refresh to reflect the changes.  

This PR fixes #9781

### CODE REVIEW

**Comment by MaddyDev** [comment]
Working on a test for this and will post it in a separate PR.

**Comment by MaddyDev** [comment]
Added debouncing to the initialize call on watch.

debouncingCallsOnEdit

File: extensions/notebook/src/book/bookTreeView.ts [hunk 1]
```diff
@@ -141,6 +140,13 @@ export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeIte
 		}
 	}
 
+	@debounce(1500)
+	async fireBookRefresh(book: BookModel): Promise<void> {
+		await book.initializeContents().then(() => {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/book/bookTreeView.ts:145]: There's still a slight possibility of a race condition here where :

1. File change causes the method to be queued up through debounce. 
1. 1.5seconds later the actual function is called. 
1. Immediately after that another file change triggers another call to be queued up. 
1. Meanwhile the first function call is taking an abnormally long time for some reason (maybe just a large book being loaded). 
1. This takes over 1.5 seconds so now the second call ends up firing and now we have two book.initializeContents calls being done at the same time which is likely to break something.

I think this is an edge case enough that we don't necessarily have to fix it now - although it wouldn't hurt to do some testing with really large books to see if there's any obvious issues we hit.

If you did want to fix it though the probable correct solution would be to capture the promise from initializeContents and then when the function is called by debounce it'd await on that promise first before firing (if the first call had already finished the promise would be resolved and immediately return).

Comment (MaddyDev) [reply extensions/notebook/src/book/bookTreeView.ts:145]: I'll test it out and add the await on promise, but I'll get this in and work on it in a different PR. Thanks Charles.

**Pull request merged**